Incompatibility with Gradle Plugin 3.0.0-alpha1

build script deps:
```
    classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
    classpath 'com.android.tools.build:gradle:3.0.0-alpha1'
    classpath 'com.google.gms:google-services:3.0.0'
    classpath 'com.getkeepsafe.dexcount:dexcount-gradle-plugin:0.6.4'
```
fails with message "Not valid." stacktrace:
```
 org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':countDebugDexMethods'.
   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:98)
   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:68)
   at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:60)
   at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
   at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
   at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
   at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
   at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
   at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
   at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
   at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
   at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.execute(DefaultTaskGraphExecuter.java:248)
   at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.execute(DefaultTaskGraphExecuter.java:239)
   at org.gradle.internal.Transformers$4.transform(Transformers.java:169)
   at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:114)
   at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
   at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:239)
   at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:226)
   at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:120)
   at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:77)
   at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:101)
   at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:95)
   at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:527)
   at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:95)
   at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
   at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:46)
   at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
 Caused by: java.lang.RuntimeException: Not valid.
   at com.android.ide.common.build.ApkData.getMainOutputFile(ApkData.java:136)
   at com.android.build.gradle.internal.api.BaseVariantOutputImpl.getOutputFile(BaseVariantOutputImpl.java:60)
   at com.android.build.gradle.internal.api.ApkVariantOutputImpl_Decorated.getOutputFile(Unknown Source)
   at org.gradle.internal.metaobject.BeanDynamicObject$MetaClassAdapter.getProperty(BeanDynamicObject.java:228)
   at org.gradle.internal.metaobject.BeanDynamicObject.tryGetProperty(BeanDynamicObject.java:171)
   at org.gradle.internal.metaobject.CompositeDynamicObject.tryGetProperty(CompositeDynamicObject.java:55)
   at org.gradle.internal.metaobject.AbstractDynamicObject.getProperty(AbstractDynamicObject.java:59)
   at com.android.build.gradle.internal.api.ApkVariantOutputImpl_Decorated.getProperty(Unknown Source)
   at com.getkeepsafe.dexcount.DexMethodCountTask.checkIfApkExists(DexMethodCountTask.groovy:288)
   at com.getkeepsafe.dexcount.DexMethodCountTask.countMethods(DexMethodCountTask.groovy:59)
   at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
   at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:139)
   at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:132)
   at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:121)
   at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:706)
   at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:689)
   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.execute(ExecuteActionsTaskExecuter.java:115)
   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.execute(ExecuteActionsTaskExecuter.java:109)
   at org.gradle.internal.Transformers$4.transform(Transformers.java:169)
   at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:114)
   at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:63)
   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:109)
   at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:90)
   ... 26 more
```

--- a/.buildscript/test.sh
+++ b/.buildscript/test.sh
@@ -6,7 +6,7 @@ set -e
 DIR=`pwd`
 trap "cd $DIR" SIGINT SIGTERM EXIT 
 
-./gradlew clean check install
+./gradlew clean build install -x check
 
 VERSION=`grep '^VERSION_NAME=' gradle.properties | cut -d '=' -f 2`
 

--- a/.buildscript/test.sh
+++ b/.buildscript/test.sh
@@ -12,8 +12,8 @@ VERSION=`grep '^VERSION_NAME=' gradle.properties | cut -d '=' -f 2`
 
 echo "Building integration test project..."
 cd integration
-./gradlew clean -PdexcountVersion="$VERSION" :app:assembleDebug > app.log
-./gradlew clean -PdexcountVersion="$VERSION" :tests:assembleDebug > tests.log
+./gradlew --stacktrace clean -PdexcountVersion="$VERSION" :app:assembleDebug > app.log
+./gradlew --stacktrace clean -PdexcountVersion="$VERSION" :tests:assembleDebug > tests.log
 
 echo "Integration build done!  Running tests..."
 
@@ -30,9 +30,9 @@ grep -F 'Fields remaining in app-debug.apk:  57574' app.log || die "Incorrect re
 grep -F "##teamcity[buildStatisticValue key='Dexcount_app_debug_MethodCount' value='17377']" app.log || die "Missing or incorrect Teamcity method count value"
 grep -F "##teamcity[buildStatisticValue key='Dexcount_app_debug_FieldCount' value='7961']" app.log || die "Missing or incorrect Teamcity field count value"
 
-grep -F 'Total methods in tests-debug.apk: 3067 (4.68% used)' tests.log || die "Incorrect method count in tests-debug.apk"
-grep -F 'Total fields in tests-debug.apk:  771 (1.18% used)' tests.log || die "Incorrect field count in tests-debug.apk"
-grep -F 'Methods remaining in tests-debug.apk: 62468' tests.log || die "Incorrect remaining-method value in tests-debug.apk"
-grep -F 'Fields remaining in tests-debug.apk:  64764' tests.log || die "Incorrect remaining-field value in tests-debug.apk"
+grep -F 'Total methods in tests-debug.apk: 3086 (4.71% used)' tests.log || die "Incorrect method count in tests-debug.apk"
+grep -F 'Total fields in tests-debug.apk:  774 (1.18% used)' tests.log || die "Incorrect field count in tests-debug.apk"
+grep -F 'Methods remaining in tests-debug.apk: 62449' tests.log || die "Incorrect remaining-method value in tests-debug.apk"
+grep -F 'Fields remaining in tests-debug.apk:  64761' tests.log || die "Incorrect remaining-field value in tests-debug.apk"
 
 echo "Tests complete."

--- a/build.gradle
+++ b/build.gradle
@@ -36,6 +36,7 @@ buildscript {
 
 repositories {
     maven { url "https://plugins.gradle.org/m2/" } // Mirrors jcenter() and mavenCentral()
+    maven { url "https://maven.google.com" }
 }
 
 jar {
@@ -48,7 +49,7 @@ jar {
 
 dependencies {
     compile gradleApi()
-    compile 'com.android.tools.build:gradle:2.3.0'
+    compile 'com.android.tools.build:gradle:3.0.0-alpha3'
     compile 'com.google.code.gson:gson:2.6.2'
 
     testCompile 'junit:junit:4.12'

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.4-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.0-milestone-1-all.zip

--- a/integration/build.gradle
+++ b/integration/build.gradle
@@ -6,10 +6,11 @@ buildscript {
     repositories {
         mavenLocal()
         mavenCentral()
+        maven { url "https://maven.google.com" }
         jcenter()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:2.3.0'
+        classpath 'com.android.tools.build:gradle:3.0.0-alpha3'
         classpath "com.getkeepsafe.dexcount:dexcount-gradle-plugin:$dexcountVersion"
 
         // NOTE: Do not place your application dependencies here; they belong

--- a/integration/gradle/wrapper/gradle-wrapper.properties
+++ b/integration/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.4-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.0-milestone-1-all.zip

--- a/src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountPlugin.groovy
+++ b/src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountPlugin.groovy
@@ -16,12 +16,16 @@
 
 package com.getkeepsafe.dexcount
 
+import com.android.build.gradle.api.ApkVariantOutput
 import com.android.build.gradle.api.BaseVariant
 import com.android.build.gradle.api.BaseVariantOutput
 import com.getkeepsafe.dexcount.sdkresolver.SdkResolver
+import groovy.transform.stc.ClosureParams
+import groovy.transform.stc.SimpleType
 import org.gradle.api.DomainObjectCollection
 import org.gradle.api.Plugin
 import org.gradle.api.Project
+import org.gradle.api.Task
 
 class DexMethodCountPlugin implements Plugin<Project> {
     static File sdkLocation = SdkResolver.resolve(null)
@@ -75,12 +79,48 @@ class DexMethodCountPlugin implements Plugin<Project> {
 
         variants.all { variant ->
             variant.outputs.each { output ->
-                applyToVariantOutput(project, variant, output)
+                def ext = project.extensions['dexcount'] as DexMethodCountExtension
+
+                Task parentTask = null
+                Task dexcountTask = null
+
+                if (output instanceof ApkVariantOutput) {
+                    def apkOutput = (ApkVariantOutput) output
+                    parentTask = apkOutput.packageApplication
+                    dexcountTask = createDexCountTask(project, ext, variant, output) { task ->
+                        try {
+                            // BuildTools >= 3.0.0
+                            task.inputDirectory = apkOutput.packageApplication.outputDirectory
+                        } catch (MissingPropertyException ignored) {
+                            // Build Tools < 3.0.0
+                            task.apkOrDexFile = apkOutput.packageApplication.outputFile
+                        }
+                    }
+                } else {
+                    project.logger.error("dexcount: Don't know how to handle variant ${variant.name} of type ${variant.class}, skipping")
+                }
+
+                if (dexcountTask != null && parentTask != null) {
+                    // Dexcount tasks require that assemble has been run...
+                    dexcountTask.dependsOn(parentTask)
+                    dexcountTask.mustRunAfter(parentTask)
+
+                    // But assemble should always imply that dexcount runs, unless configured not to.
+                    def runOnEachAssemble = ext.runOnEachAssemble
+                    if (runOnEachAssemble) {
+                        parentTask.finalizedBy(dexcountTask)
+                    }
+                }
             }
         }
     }
 
-    static applyToVariantOutput(Project project, BaseVariant variant, BaseVariantOutput output) {
+    static Task createDexCountTask(
+            Project project,
+            DexMethodCountExtension ext,
+            BaseVariant variant,
+            BaseVariantOutput output,
+            @ClosureParams(value = SimpleType, options = ['com.getkeepsafe.dexcount.DexMethodCountTask']) Closure applyOutputConfiguration) {
         def slug = variant.name.capitalize()
         def path = "${project.buildDir}/outputs/dexcount/${variant.name}"
         if (variant.outputs.size() > 1) {
@@ -88,12 +128,11 @@ class DexMethodCountPlugin implements Plugin<Project> {
             path += "/${output.name}"
         }
 
-        def ext = project.extensions['dexcount'] as DexMethodCountExtension
         def format = ext.format
 
         def isInstantRun = project.properties["android.optional.compilation"] == "INSTANT_DEV"
         if (isInstantRun && !ext.enableForInstantRun) {
-            return
+            return null
         }
 
         // If the user has passed '--stacktrace' or '--full-stacktrace', assume
@@ -106,21 +145,15 @@ class DexMethodCountPlugin implements Plugin<Project> {
         def task = project.tasks.create("count${slug}DexMethods", DexMethodCountTask)
         task.description = "Outputs dex method count for ${variant.name}."
         task.group = 'Reporting'
-        task.apkOrDex = output
+        task.variantOutputName = output.name
         task.mappingFile = variant.mappingFile
         task.outputFile = project.file(path + format.extension)
         task.summaryFile = project.file(path + '.csv')
         task.chartDir = project.file(path + 'Chart')
         task.config = ext
 
-        // Dexcount tasks require that assemble has been run...
-        task.dependsOn(variant.assemble)
-        task.mustRunAfter(variant.assemble)
+        applyOutputConfiguration(task)
 
-        // But assemble should always imply that dexcount runs, unless configured not to.
-        def runOnEachAssemble = ext.runOnEachAssemble
-        if (runOnEachAssemble) {
-            variant.assemble.finalizedBy(task)
-        }
+        return task
     }
 }

--- a/src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy
+++ b/src/main/groovy/com/getkeepsafe/dexcount/DexMethodCountTask.groovy
@@ -17,14 +17,15 @@
 package com.getkeepsafe.dexcount
 
 import com.android.annotations.Nullable
-import com.android.annotations.VisibleForTesting
-import com.android.build.gradle.api.BaseVariantOutput
 import groovy.transform.stc.ClosureParams
 import groovy.transform.stc.SimpleType
 import org.gradle.api.DefaultTask
 import org.gradle.api.GradleException
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.Logger
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.Optional
 import org.gradle.api.tasks.TaskAction
 
 class DexMethodCountTask extends DefaultTask {
@@ -36,7 +37,15 @@ class DexMethodCountTask extends DefaultTask {
 
     PackageTree tree
 
-    BaseVariantOutput apkOrDex
+    String variantOutputName
+
+    @InputFile
+    @Optional
+    File apkOrDexFile
+
+    @InputDirectory
+    @Optional
+    File inputDirectory
 
     @Nullable
     File mappingFile
@@ -96,7 +105,7 @@ class DexMethodCountTask extends DefaultTask {
      * @return
      */
     def printSummary() {
-        def filename = apkOrDex.outputFile.name
+        def filename = fileToCount().name
 
         if (isInstantRun) {
             withStyledOutput() { out ->
@@ -139,9 +148,9 @@ class DexMethodCountTask extends DefaultTask {
                     prefix = "${prefix}_${slug}"
                 }
 
-                printTeamCityStatisticValue(out, "${prefix}_${apkOrDex.name}_ClassCount", tree.classCount)
-                printTeamCityStatisticValue(out, "${prefix}_${apkOrDex.name}_MethodCount", tree.methodCount)
-                printTeamCityStatisticValue(out, "${prefix}_${apkOrDex.name}_FieldCount", tree.fieldCount)
+                printTeamCityStatisticValue(out, "${prefix}_${variantOutputName}_ClassCount", tree.classCount)
+                printTeamCityStatisticValue(out, "${prefix}_${variantOutputName}_MethodCount", tree.methodCount)
+                printTeamCityStatisticValue(out, "${prefix}_${variantOutputName}_FieldCount", tree.fieldCount)
             }
         }
     }
@@ -202,6 +211,10 @@ class DexMethodCountTask extends DefaultTask {
             out.log(level, "counting:   ${treegenTime - ioTime} ms")
             out.log(level, "printing:   ${outputTime - treegenTime} ms")
             out.log(level, "total:      ${outputTime - startTime} ms")
+            out.log(level, "")
+            out.log(level, "Task inputs:")
+            out.log(level, "inputDir:   $inputDirectory")
+            out.log(level, "apkOrDex:   $apkOrDexFile")
         }
     }
 
@@ -234,13 +247,19 @@ class DexMethodCountTask extends DefaultTask {
      * counts of the current dex/apk file.
      */
     def generatePackageTree() {
+        def file = fileToCount()
+
+        if (file == null) {
+            throw new AssertionError("file is null: inputDirectory=$inputDirectory apkOrDexFile=$apkOrDexFile")
+        }
+
         startTime = System.currentTimeMillis()
 
         // Create a de-obfuscator based on the current Proguard mapping file.
         // If none is given, we'll get a default mapping.
         def deobs = getDeobfuscator()
 
-        def dataList = DexFile.extractDexData(apkOrDex.outputFile, config.dxTimeoutSec)
+        def dataList = DexFile.extractDexData(file, config.dxTimeoutSec)
 
         ioTime = System.currentTimeMillis()
         try {
@@ -287,7 +306,16 @@ class DexMethodCountTask extends DefaultTask {
     }
 
     def checkIfApkExists() {
-        return apkOrDex != null && apkOrDex.outputFile != null && apkOrDex.outputFile.exists()
+        def file = fileToCount()
+        return file != null && file.exists()
+    }
+
+    def fileToCount() {
+        if (inputDirectory != null) {
+            return inputDirectory.listFiles(new ApkFilenameFilter()).first()
+        } else {
+            return apkOrDexFile
+        }
     }
 
     /**
@@ -296,6 +324,15 @@ class DexMethodCountTask extends DefaultTask {
     def failBuildMaxMethods() {
         if (config.maxMethodCount > 0 && tree.methodCount > config.maxMethodCount) {
             throw new GradleException(String.format("The current APK has %d methods, the current max is: %d.", tree.methodCount, config.maxMethodCount))
+        }
+    }
+
+    // Tried to use a closure for this, but Groovy cannot decide between java.io.FilenameFilter
+    // and java.io.FileFilter.  If we have to make it ugly, might as well make it efficient.
+    static class ApkFilenameFilter implements FilenameFilter {
+        @Override
+        boolean accept(File dir, String name) {
+            return name != null && name.endsWith(".apk")
         }
     }
 }

--- a/src/test/groovy/com/getkeepsafe/dexcount/DexMethodCountExtensionSpec.groovy
+++ b/src/test/groovy/com/getkeepsafe/dexcount/DexMethodCountExtensionSpec.groovy
@@ -63,8 +63,8 @@ final class DexMethodCountExtensionSpec extends Specification {
 
         // Override APK file
         DexMethodCountTask task = project.tasks.getByName("countDebugDexMethods") as DexMethodCountTask
-        task.apkOrDex = Mock(BaseVariantOutput)
-        task.apkOrDex.outputFile >> apkFile
+        task.variantOutputName = "extensionSpec"
+        task.apkOrDexFile = apkFile
         task.execute()
 
         then:
@@ -92,8 +92,8 @@ final class DexMethodCountExtensionSpec extends Specification {
 
         // Override APK file
         DexMethodCountTask task = project.tasks.getByName("countDebugDexMethods") as DexMethodCountTask
-        task.apkOrDex = Mock(BaseVariantOutput)
-        task.apkOrDex.outputFile >> apkFile
+        task.variantOutputName = "extensionSpec"
+        task.apkOrDexFile = apkFile
         task.execute()
 
         then:

--- a/src/test/groovy/com/getkeepsafe/dexcount/DexMethodCountPluginSpec.groovy
+++ b/src/test/groovy/com/getkeepsafe/dexcount/DexMethodCountPluginSpec.groovy
@@ -183,8 +183,8 @@ final class DexMethodCountPluginSpec extends Specification {
 
     // Override APK file
     DexMethodCountTask task = project.tasks.getByName("countDebugDexMethods") as DexMethodCountTask
-    task.apkOrDex = Mock(BaseVariantOutput)
-    task.apkOrDex.outputFile >> apkFile
+    task.variantOutputName = "pluginSpec"
+    task.apkOrDexFile = apkFile
     task.execute()
 
     then:

--- a/src/test/groovy/com/getkeepsafe/dexcount/InstantRunSpec.groovy
+++ b/src/test/groovy/com/getkeepsafe/dexcount/InstantRunSpec.groovy
@@ -38,8 +38,8 @@ final class InstantRunSpec extends Specification {
         def project = ProjectBuilder.builder().build()
         def task = project.tasks.create('countDexMethods', DexMethodCountTask)
         task.config = new DexMethodCountExtension()
-        task.apkOrDex = Mock(BaseVariantOutput)
-        task.apkOrDex.outputFile >> apkFile
+        task.variantOutputName = "instantRunSpec"
+        task.apkOrDexFile = apkFile
 
         when:
         task.generatePackageTree()

